protocol/messenger_communities.go

```diff
@@ -503,16 +504,38 @@ func (m *Messenger) startCuratedCommunitiesUpdateLoop() {
 	// The current contract contains communities that are no longer accessible on waku
 	const unknownCommunitiesFoundTimeout = 60 * time.Second
 
+	type curatedCommunities struct {
+		ContractCommunities         []string
+		ContractFeaturedCommunities []string
+		UnknownCommunities          []string
+	}
+
+	var mu = sync.RWMutex{}
+	var c = curatedCommunities{}
+
 	go func() {
 		for {
 			var timeTillNextUpdate time.Duration
 
 			response, err := m.CuratedCommunities()
+
 			if err != nil {
 				timeTillNextUpdate = errorTimeout
 			} else {
-
-				m.config.messengerSignalsHandler.SendCuratedCommunitiesUpdate(response)
+				// Check if it's the same values we had
+				if !reflect.DeepEqual(c.ContractCommunities, response.ContractCommunities) ||
+					!reflect.DeepEqual(c.ContractFeaturedCommunities, response.ContractFeaturedCommunities) ||
+					!reflect.DeepEqual(c.UnknownCommunities, response.UnknownCommunities) {
+					// One of the communities is different, send the updated response
+					m.config.messengerSignalsHandler.SendCuratedCommunitiesUpdate(response)
+
+					// Update the values
+					mu.Lock()
+					c.ContractCommunities = response.ContractCommunities
+					c.ContractFeaturedCommunities = response.ContractFeaturedCommunities
+					c.UnknownCommunities = response.UnknownCommunities
+					mu.Unlock()
```

Comment:
The check-and-update mechanism for the `c` variable is not atomic, which might lead to race conditions if there are other parts of the code concurrently reading/writing to `c`. You're using sync.RWMutex, but you are only locking for the update part.

Reply:
I only put it around the update part since `RWMutex` enables multiple readers, but just to not take chances, I put the lock around the read as well now.